README.rst
Flask restful service in docker
===============================

**GitHub repo**: <https://github.com/And0k/flask_rest>

**Docker Hub image**: <https://hub.docker.com/r/and0k/flask_rest>


Description
-----------

Python Flask restful API running with `Meinheld <https://github.com/mopemope/meinheld>`_ controlled by `Gunicorn <http://gunicorn.org>`_.
When receives list of urls of images it converts them to black and white and serve in local folder.

Test restful service without docker
-----------------------------------

* run ``pipenv install -dev``
* run tests using pipenv from ``/test/``
* to run Flask development server run ``manage.py`` from ``/service_images/`` directory.

Docker image was build on Alpine Linux and has size 155MB. Download image: ``docker pull and0k/flask_rest``

Build docker image on Linux
---------------------------

Clone repository and run from its directory:

``pipenv install``

``su``

``systemctl start docker``

``docker-compose -f docker-compose.yml up -d``

Try out JSON and HTML requests
------------------------------
API supports content negotiation so you can browse results starting from main page ``http://127.0.0.1:5000/``.
Send requests directly to this address or REST API using documentation provided by `Swagger UI <https://swagger.io/tools/swagger-ui/>`_ at
``http://127.0.0.1:5000/doc``.

---

This is a test. Original task (russian) see in ``/system-tech_task[ru].rst``


ChangeLog
CHANGES
=======

* remove garbage
* Initial commit
* flask\_rest on meinheld controlled by gunicorn converts images in alpine docker


test/config.py
from pathlib import Path
import logging

class Config:
    TESTING = True
    TEST_PATH = Path(__file__).parent
    PROCESSED_DIR_PATH = TEST_PATH / 'images_out'
    LOG_PATH = TEST_PATH / 'server.log'
    LOGGING_LEVEL = logging.DEBUG


test/conftest.py
import pytest
from .config import Config
from ..service_images import create_app


#@pytest.fixture(scope='module')
# def create_app():
#     return a  # loads test config


@pytest.fixture(scope='module')
def app():
    """
    fixture whose name is "app"
    Adds config parameters for test
    :return: flask server instance
    """

    a = create_app(Config)
    #from ..service_images import apis
    return a


@pytest.fixture(scope='module')
def testdata_dir(app):
    """
    Info about data dirs for test
    :param parent_path: :class:`pathlib.Path`, parent dir of test data
    :return: function turned to dict value if used in test
    """
    # if not guid in {dir.name for dir in list_result_dirs()}:  # prepared?
    #     guid = list_result_dirs()[0].name
    #     print('Selected 1st dir as guid')

    test_data_dir_dict = {
        'images_in': app.config['TEST_PATH'] / 'data' / 'in' / 'images_in',
        'guids_in': app.config['TEST_PATH'] / 'data' / 'in',
        'images_out': app.config['TEST_PATH'] / 'data' / 'images_out'

    }

    def dict_value(key):
        return test_data_dir_dict[key]

    return dict_value


@pytest.fixture
def processed_dir_path_with_guids(testdata_dir, app):
    """
    Temporary point app.config['PROCESSED_DIR_PATH'] to test_data_dir_dict['guids_in']
    :param app:
    :return: app.config['PROCESSED_DIR_PATH']
    """

    save_path = app.config['PROCESSED_DIR_PATH']
    app.config['PROCESSED_DIR_PATH'] = testdata_dir('guids_in')
    yield testdata_dir('guids_in')
    app.config['PROCESSED_DIR_PATH'] = save_path


@pytest.fixture
def tmp_processed_dir_path(tmp_path, app):
    """
    Temporary make new temporary dir for app processing results and point app.config['PROCESSED_DIR_PATH'] to it
    :param tmp_path:
    :param app:
    :return:
    """
    save_path = app.config['PROCESSED_DIR_PATH']
    app.config['PROCESSED_DIR_PATH'] = tmp_path
    yield tmp_path
    app.config['PROCESSED_DIR_PATH'] = save_path


@pytest.fixture  #(scope='module')
def list_result_dirs(app):
    def list_result_dirs_func():
        return [d for d in app.config['PROCESSED_DIR_PATH'].iterdir() if d.is_dir()]
    return list_result_dirs_func
    # return [d for d in app.config['PROCESSED_DIR_PATH'].iterdir() if d.is_dir()]
    # def list_result_dirs_func(dir=(Path(dir) if dir else tmp_path)):
    #     return [d for d in dir.iterdir() if d.is_dir()]
    # return list_result_dirs_func




# @pytest.fixture
# def rem_dir_contents(dir):
#     # prepare (i.e. empty) result dir from subdirs
#     for d in list_result_dirs(dir):
#         # remove all files from dir and then dir itself
#         for file in d.iterdir():
#             file.unlink()
#         d.rmdir()
#     assert len(list_result_dirs(dir)) == 0  # prepared?


def pytest_report_header(config):
    return "mailto: ao.korzh@gmail.com"

test/func_test.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
"""
  Purpose:  functional tests
  Author:   Andrey Korzh <ao.korzh@gmail.com>
  Created:  25.01.2019
"""
import pytest
from flask import url_for
from PIL import Image
from io import BytesIO
from pathlib import Path

url_images = ["https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png",
              "http://flask.pocoo.org/docs/1.0/_images/logo-full.png"]  # if copy to json use _"_ not _'_
len_hex_uuid4 = 32
tmp_dir_name = f'{"~test_id_to_delete~":~<{len_hex_uuid4}}'

#@pytest.mark.skip(reason="not now")
def test_post_task(client, tmp_processed_dir_path, list_result_dirs):
    """
    http post: send urls for processing
    check: images from urls are processed and saved in new folder named with generated GUID which is returned
    :param client: fixture, injected by the `pytest-flask` plugin
    :param tmp_processed_dir_path: fixture, using it ensures we have new temporary dir
    :return:
    """

    # query
    n_urls = len(url_images)
    response = client.post("/", json = {'url_list': url_images, 'threshold': 150})

    # Was 1 dir created?
    assert len(list_result_dirs()) == 1

    # Was it filled with files number equal number of queried urls?
    assert len([1 for _ in list_result_dirs()[0].iterdir()]) == n_urls

    # Validate the response
    assert response.status_code == 201
    assert response.json == {
        'guid': list_result_dirs()[0].name,
        'processed_cnt': n_urls
    }


@pytest.fixture
def guids(list_result_dirs):
    """
    work folders from file system (autonamed folders has guid names)
    :param list_result_dirs: fixture
    :return:
    """
    ids = [d.name for d in list_result_dirs()]
    assert len(ids)>0  # test folder prepared?
    return ids


@pytest.mark.parametrize('url', ['', '/', '/index', '/index/'])
def test_get_json_ids_list(client, url, processed_dir_path_with_guids, guids):
    """
    http get main page (correspond to any of 4 parametrized urls)
    receive same list of work folders in json
    :param client: fixture, injected by the `pytest-flask` plugin
    :param processed_dir_path_with_guids: fixture, using it ensures we have prepared dir (where exist folders are guids)
    :return:
    """
    print(f'get folders from "{str(processed_dir_path_with_guids)}" for url "{url}"')
    response = client.get(url)
    # receive json with work ids?
    assert response.status_code == 200
    assert response.json == {'guids': guids}


id = '5e0ca894f29d4395ba234e1fe865ecea'  # folder we check, must be in _data/in_ folder
i_file = 1   #index of file we check


def test_get_json_files_list(client, app, list_result_dirs, processed_dir_path_with_guids, testdata_dir, guids):
    """
    http get _address of urls list_ of (processed) files for one of ids
    receive list with contents of corresponding folder given by urls relative to address for Config['PROCESSED_DIR_PATH']
    :param processed_dir_path_with_guids: fixture, using it ensures we have prepared dir (where exist folders are guids)
    :param guids: work folders red from file system
    :global param: id
    """

    url = f"/{id}"
    print(f'get folders from "{str(processed_dir_path_with_guids)}" for url "{url}"')
    response = client.get(url)
    assert response.status_code == 200
    # receive json with urls of result files?
    assert response.json == {
        'files': [url_for('static', filename=str(file.relative_to(app.config['PROCESSED_DIR_PATH']))) for file in \
                  Path(testdata_dir('guids_in') / id).iterdir()
                  ]
    }


def test_get_file(client, app, processed_dir_path_with_guids):
    """
    http get address of 1 (processed) file for one of ids
    receive 1 file of Config['PROCESSED_IMAGE_FORMAT'] format
    :param processed_dir_path_with_guids: fixture, using it ensures we have prepared dir (where exist folders are guids)
    :global param: id, i_file
    """
    response = client.get(f"/{id}/{i_file}.{app.config['PROCESSED_IMAGE_FORMAT']}/")
    assert response.status_code == 200
    img = Image.open(BytesIO(response.data))
    assert img.format == app.config['PROCESSED_IMAGE_FORMAT'].upper()


@pytest.fixture(params=range(4), ids= ['""', '/', '/guid', '/guid/file']) #
def query_urls(request, app):
    check_urls = ['',
                  '/',
                  f'/{id}',
                  f"/{id}/{i_file}.{app.config['PROCESSED_IMAGE_FORMAT']}"]
    response_contains = [
        id,
        id,
        f"/{id}/{i_file}.{app.config['PROCESSED_IMAGE_FORMAT']}",
        'PNG']
    return ((check_urls[request.param], response_contains[request.param]),)


def test_get_html(client, app, query_urls, processed_dir_path_with_guids):
    """
    Check content negotiation. Http get queries with {'Accept': 'text/html'}
    receive from {'Accept': 'text/html'} queries has html result
    :param processed_dir_path_with_guids: fixture, using it ensures we have prepared dir (where exist folders are guids)
    :return:
    """
    url, response_contains = query_urls[0]
    headers = 'text/html'  # 'headers', ['text/html']
    response = client.get(url, headers={'Accept': headers})
    assert response_contains.encode() in response.data
    if url.strip('/').endswith(app.config['PROCESSED_IMAGE_FORMAT']):
        assert response.mimetype.startswith('image')
    else:
        assert response.mimetype == headers


def test_delete(client, tmp_processed_dir_path, list_result_dirs):
    """

    :param client:
    :param tmp_processed_dir_path: fixture, using it ensures we have new temporary dir
    :return:
    """
    # make some dir inside temorary dir and write file
    new_dir = tmp_processed_dir_path / tmp_dir_name
    new_dir.mkdir()
    (new_dir / 'somefile').touch()
    # Was 1 dir created?
    assert len(list_result_dirs()) == 1

    # query to delete dir
    response = client.delete(f"/{new_dir.name}")

    # Was dir deleted?
    assert len(list_result_dirs()) == 0

    # Validate the response
    assert response.status_code == 204
    # assert response.json == {
    #     'guid': list_result_dirs()[0].name,
    #     'processed_cnt': 2
    # }
    pass

def test_update_task(client, tmp_processed_dir_path, list_result_dirs):
    """
    http post: send urls for processing
    check: images from urls are processed and saved in new folder named with generated GUID which is returned
    :param client: fixture, injected by the `pytest-flask` plugin
    :param tmp_processed_dir_path: fixture, using it ensures we have new temporary dir
    :return:
    """

    # make some dir inside temorary dir and write file
    new_dir = (tmp_processed_dir_path / tmp_dir_name)
    new_dir.mkdir()
    (new_dir / 'somefile').touch()
    # Was 1 dir created?
    assert len(list_result_dirs()) == 1

    # query to update dir
    n_urls = len(url_images)
    response = client.put(f"/{new_dir.name}", json = {'url_list': url_images, 'threshold': 150})

    # same dir exists?
    assert list_result_dirs()[0] == new_dir

    # Was it filled with files number equal number of queried urls?
    assert len([1 for _ in list_result_dirs()[0].iterdir()]) == n_urls

    # Validate the response
    assert response.status_code == 201
    assert response.json == {
        'guid': list_result_dirs()[0].name,
        'processed_cnt': n_urls
    }


@pytest.mark.skip(reason="run only if need print your API as a Postman")
def postman_helper(app):
    from flask import json
    urlvars = False     # Build query strings in URLs
    swagger = True      # Export Swagger specifications
    data = app.as_postman(urlvars=urlvars, swagger=swagger)
    print(json.dumps(data))

service_images/models.py
import uuid, logging
from pathlib import Path
from io import BytesIO
import requests
from PIL import Image
from .common import ApiError
from flask import current_app as app
logger = logging.getLogger('flask.app')


def remove_transparency(im, bg_colour=(255, 255, 255)):
    """
    Remove transparency/alpha from PIL.Image
    https://stackoverflow.com/a/35859141
    :param im: PIL.Image
    :param bg_colour: tuple (R, G, B), background color will be used instead transparency
    :return:
    """
    # Only process if image has transparency (http://stackoverflow.com/a/1963146)
    if im.mode in ('RGBA', 'LA') or (im.mode == 'P' and 'transparency' in im.info):

        # Need to convert to RGBA if LA format due to a bug in PIL (http://stackoverflow.com/a/1963146)
        alpha = im.convert('RGBA').split()[-1]

        # Create a new background image of our matt color.
        # Must be RGBA because paste requires both images have the same format
        # (http://stackoverflow.com/a/8720632  and  http://stackoverflow.com/a/9459208)
        bg = Image.new("RGBA", im.size, bg_colour + (255,))
        bg.paste(im, mask=alpha)
        return bg
    else:
        return im


def process_image(file, out_name=None, threshold=200):
    """
    Converts image to black and white with given threshold
    :param out_name: pathlib.Path, name of result image file
    :param file: [str, :class:pathlib.Path] - filename to read image or
                 file object (see PIL.open())
    :param threshold: int or float, in range 0..255. If higher then result will blacker
    :return: None
    """
    if not file:
        return
    try:
        img_in = Image.open(file)
        img_in = remove_transparency(img_in)
        # convert to 8-bit pixels, black and white, with given threshold
        img_out = img_in.convert('L').point(lambda x: 255 if x > threshold else 0, mode='1')
        img_out.save(out_name)
    except Exception as err:
        raise Exception('Processing image error.') from err


def request_file(url):
    """
    http get url with timeouts, mimic header of Mozilla browser
    :param url:
    :return: file object
    """
    connect_timeout = 5
    read_timeout = 50
    headers = {
        'User-Agent': 'Mozilla/5.0'}  # may add more info: (Macintosh; Intel Mac OS X 10_9_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.47 Safari/537.36
    try:
        response = requests.get(url, headers=headers, timeout=(connect_timeout, read_timeout))
        response.raise_for_status()
    except Exception as err:  # (requests.HTTPError, requests.Timeout, ConnectionError)
        raise ApiError(str(err), details='Get image error')
    return BytesIO(response.content)


class WorkDAO(object):
    """
    Required app.config fields:
    PROCESSED_DIR_PATH
    PROCESSED_IMAGE_FORMAT
    """
    def __init__(self):
        if app.config['PROCESSED_DIR_PATH'].is_dir():
            pass  # self.ids = {d.name for d in self.list_ids()}
        else:
            app.config['PROCESSED_DIR_PATH'].mkdir(parents=True)

            # self.ids = set()  # set of str, folder names, maintained here in sync with existed folders in app.config['PROCESSED_DIR_PATH']

    def get(self, id):
        """
        Find folder with result of previous work.
        :param id: str
        :return: list of pathlib.Path, contents of folder
        Allows to not maintain folder list in memory/database
        """
        return self.find_work_dir(id).iterdir()

    def get_item(self, id, item):
        """
        :param id: str
        :param item: str
        :return: :class:`pathlib.Path` to file in dir
        """
        file = self.find_work_dir(id) / item
        return file

    def create(self, url_list, threshold, id=None):
        """
        Process all images links, saves to result dir and returns its name
        :param url_list: task data
        :param threshold: task parameter
        :param id: None to generate unique work id or str to create folder with this name
        :return:
        """
        if not id:
            # generate unique work id
            id = uuid.uuid4().hex  # will use hex representation of generated
        work_dir = app.config['PROCESSED_DIR_PATH'] / id
        try:
            work_dir.mkdir(exist_ok=True)
            processed_cnt = 0
            for i, url in enumerate(url_list):
                try:
                    file_obj = request_file(url)
                    process_image(file_obj, out_name=work_dir / f"{i}.{app.config['PROCESSED_IMAGE_FORMAT']}", threshold=threshold)
                    processed_cnt += 1
                except Exception as err:
                    logger.warning(f'{err} #{i} {url} failed')
            return {'guid': id, 'processed_cnt': processed_cnt}  # 'Images processed'
        except Exception as err:
            if Path.is_dir(work_dir):  # remove not completed work result
                Path.rmdir(work_dir)
            ApiError(str(err), f'Data not processed', 404)

    def update(self, id, url_list, threshold):
        self.delete(id)
        return self.create(url_list, threshold, id)

    def delete(self, id):
        """
        remove dir of id
        :param id: str
        :return:
        """

        dir = app.config['PROCESSED_DIR_PATH'] / id
        # remove all files from dir (required to can remove dir itself)
        for file in dir.iterdir():
            file.unlink()
        dir.rmdir()
        return dir.name

    def find_work_dir(self, id):
        """
        path to dir of id
        :param id: str, work id
        :return: pathlib.Path to dir
        """
        try:
            return app.config['PROCESSED_DIR_PATH'] / id
        except Exception as err:
            ApiError(str(err), f'Have no saved data with id {id}', 404)

    def list_ids(self):
        return [d.name for d in app.config['PROCESSED_DIR_PATH'].iterdir() if d.is_dir()]



service_images/__init__.py
from flask import Flask, got_request_exception, render_template, json
from flask.logging import default_handler
import logging
from .config import Config
# from .models import WorkDAO
from flask_restplus import Api as BaseApi


# configure all loggers same as default flask logger
root = logging.getLogger()
root.addHandler(default_handler)


def create_app(config_class=None):
    """
    App factory
    :param config_class: class, with additional config parameters if need
    :return: Flask instance
    """
    app = Flask(__name__,
                static_url_path = ''  # removes any preceding path from the URL (i.e. the default /static)
                ) # template_folder = 'my/templates'

    app.url_map.strict_slashes = False
    app.config.from_object(Config)          # default
    app.config.from_object(config_class)    # overwrite & additional

    app.static_folder = str(app.config['PROCESSED_DIR_PATH'].resolve())

    # Add file logger for the app
    file_handler = logging.FileHandler(filename=app.config['LOG_PATH'])
    file_handler.setFormatter(logging.Formatter('%(asctime)s\t%(levelname)s\t%(message)s'))
    app.logger.addHandler(file_handler)
    app.logger.setLevel(app.config['LOGGING_LEVEL'])

    # Attach custom error handlers to an exception
    def log_exception(sender, exception, **extra):
        """ Log an exception to our logging framework """
        sender.logger.debug('Got exception during processing: %s', exception)

    got_request_exception.connect(log_exception, app)


    # Hack to use current configuration
    with app.app_context():
        from .apis import api as ns1

    # Workaround of "I can't seem to set a path of '/'": https://github.com/noirbizarre/flask-restplus/issues/247
    # - solution 1
    class Api(BaseApi):
        def _register_doc(self, app_or_blueprint):
            # HINT: This is just a copy of the original implementation but with the last line commented out.
            if self._add_specs and self._doc:
                # Register documentation before root if enabled
                app_or_blueprint.add_url_rule(self._doc, 'doc', self.render_doc)
            # app_or_blueprint.add_url_rule(self._doc, 'root', self.render_root)

        @property
        def base_path(self):
            return ''

    api = Api(
        title='Image processing service',
        version='1.0',
        description='Test task of implementing The REST API',
        doc = '/doc/'
    )
    app.config.SWAGGER_UI_DOC_EXPANSION = 'list'

    from flask import make_response, render_template_string

    def convert_dict_of_lists_to_html(data):
        """
        Render lists from each key, also render images if have corresponded key
        :param data: dict with field(s) which contain list
        :return:
        """
        return render_template_string(""" 
                        {% for key, list in data.items() %}
                           <h2> {{key}} </h2>
                           <ul>
                           {% if key != 'files' %}
                                {% for item in list %}
                                  <li><a href = "{{ item }}">{{ item }}</a> </li>
                                {% endfor %}
                           {% else %}
                                {% for file in list %}
                                  <li>
                                    <a href = "{{ file }}">{{ file }}</a>
                                    <a download="{{ file }}" href="{{ file }}" title="{{ file }}">
                                      <img src="{{ file }}" width="120" height="90" />
                                    </a>
                                  </li>
                                {% endfor %}
                           {% endif %}                                    
                           </ul>
                        {% endfor %}
                        """, data=data)

    @api.representation('text/html')
    def output_html(data, code, headers=None):
        resp = make_response(convert_dict_of_lists_to_html(data), code)
        if headers:
            resp.headers.extend(headers)
        return resp

    @api.representation('application/json')
    def output_json(data, code, headers=None):
        resp = make_response(json.dumps(data), code)
        if headers:
            resp.headers.extend(headers)
        return resp

    @api.errorhandler
    def default_error_handler(error):
        """
        Default error handler
        :param error:
        :return:
        """
        app.logger.exception(error)
        return {'message': str(error)}, getattr(error, 'code', 500)

    # - solution 2
    # @app.route('/')
    # def index():
    #     return render_template('index.html')


    api.add_namespace(ns1)
    api.init_app(app)  # app.register_blueprint(api, url_prefix='/api/1')
    return app



service_images/apis.py
from flask import current_app as app, request, jsonify, make_response, send_file, send_from_directory, render_template_string, url_for
from flask_restplus import Namespace, Resource, fields, marshal_with, reqparse, marshal
from pathlib import Path
from .models import WorkDAO, ApiError
# from werkzeug.​contrib.​fixers import ProxyFix

#blueprint = Blueprint('api', __name__, url_prefix='/api')
api = Namespace('Tasks', path='/', title='Images processing API', version='1.0', contact='ao.korzh@gmail.ru',
          description='A simple API of converting images to black and white', validate=True)
Work = WorkDAO()  #processed_dir_path=app.config['PROCESSED_DIR_PATH']

# api.namespaces.clear()
# ns = api.namespace('tasks', description='operations')

# def on_json_loading_failed(err):
#     if not request.is_json:
#         raise ApiError('Expected "application/json" encoded data', 'Wrong media type', 415)
#     raise ApiError(str(err), 'Bad request', 400)
#
#
# request.on_json_loading_failed = on_json_loading_failed
# task = request.get_json(
#     force=True)  # force=True will make sure this works even if a client does not specify application/json

# input models
len_hex_uuid4 = 32
guid_model_field = fields.String(description='The task unique identifier (hex uuid4 number without dashes)',
                          min_length=len_hex_uuid4, max_length=len_hex_uuid4)
task_model_dict = {
    'url_list': fields.List(fields.Url, description='The task processing parameter: list of urls of images to process'),
    'threshold': fields.Float(required=True, min=0, max=255, description='The task processing parameter')
}
task_model = api.model('task', task_model_dict)
#put_task_fields = api.model('put task', dict(post_task_model_dict, guid=guid_model_field))

# output model
processed_fields = api.model('Processed', {
    'guid': guid_model_field,
    'processed_cnt': fields.Integer})
# class Processed(fields.Raw):
#     def format(self, value):
#         return {'id': value.id, 'processed_cnt': value.processed_cnt}


# Resources #
@api.route('/', '/index/')        #, endpoint='root'
class ProcessedTasks(Resource):         # Root
    # def __init__(self):
    #     parser = reqparse.RequestParser()
    #     # Look only in the json
    #     parser.add_argument('url_list', type=list, location='json')
    #     parser.add_argument('threshold', type=list, location='json')


    def get(self):
        """
        List all work id (folders named with GUIDs if created by Post message)
        :return: list of all ids
        """
        return {'guids': Work.list_ids()}


    @api.expect(task_model)
    @marshal_with(processed_fields)
    def post(self):
        """
        Data processing request.
        Input json data are url list of links to images
        :return: json with guid of unique work ID (and the same name of folder with results and number of successfully processed files)
        """
        #args = self.reqparse.parse_args()
        return Work.create(**api.payload), 201


# @api.​response(404, 'Task id not found')
# @api.​param('guid', 'The task identifier')
@api.route('/<string:guid>/<path:filename>/')   # @api.route('/<path:path>')   #
class ProcessedTaskItem(Resource):
    """

    """

    def get(self, guid, filename=None): #
        """
        :param guid:
        :param filename:
        :return: individual image
        """
        if filename:
            file = Work.get_item(guid, filename)
            try:
                return send_from_directory(
                    app.config['PROCESSED_DIR_PATH'] / guid, str(file.name), as_attachment=True,
                    attachment_filename=filename)
            except Exception as err:
                raise ApiError(f'{filename} not in dir', 'Bad request', 400)
                # api.abort(http_code, self.message)
        # else:
        #     raise ApiError(f'bad format', 'Bad request', 400)

@api.route(f'/<string(length={len_hex_uuid4}):guid>/')  #, doc={'params':task_fields['guid']} @api.route('/<path:path>')   # )
class ProcessedTaskItems(Resource):
    """

    """

    # @api.doc('return processed image')
    # @api.expect(task_fields, validate=True)
    # @api.marshal_with(task_fields, code=201)
    # @api.representation('image/png')
    # def __init__(self, filename):
    #     """
    #     Split filename on work id and
    #     :param filename:
    #     """
    #     parser = reqparse.RequestParser()
    #     # Look only in the json
    #     parser.add_argument('url_list')
    #     parser.add_argument('threshold')
    #     self.guid, self.filename = filename.split('/', 1)

    #@api.marshal_list_with({'files': fields.List(fields.String, description='image file name with extention')})
    def get(self, guid=None):  #
        """
        url list of processed files
        :param guid: globally unique work id
        :return:
            json: processed image(s) urls or
            html: all in html list
        """
        # All images in list
        files = list(Work.get(guid))
        return {'files':
                    [url_for('static',
                             filename=str(file.relative_to(app.config['PROCESSED_DIR_PATH']))) for file in files]
               }

    #@api.doc('delete processed task dir')
    @api.response(204, "{'result': True}" )  #'Processed task deleted'
    def delete(self, guid):
        """Delete a processed task results given its guid"""
        if Work.delete(guid):
            return {'result': True}, 204   # {'deleted': guid}, 204

    @api.expect(task_model)
    def put(self, guid):
        """Update a processed task results given its identifier"""
        return Work.update(guid, **api.payload), 201
